src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,571 @@
+    event File(bytes32 indexed what, address data);
+    event Wind(bytes32 indexed ilk, uint256 amount);
+    event Unwind(bytes32 indexed ilk, uint256 amount, uint256 fees);
+    event Reap(bytes32 indexed ilk, uint256 amt);
+    event Cage();
+    event Cage(bytes32 indexed ilk);
+    event Cull(bytes32 indexed ilk);
+    event Uncull(bytes32 indexed ilk);
+    event Quit(bytes32 indexed ilk, address indexed usr);
+    event Exit(bytes32 indexed ilk, address indexed usr, uint256 amt);
+
+    /**
+        @dev sets msg.sender as authed.
+        @param vat_     address of the DSS vat contract
+        @param daiJoin_ address of the DSS Dai Join contract
+    */
+    constructor(address vat_, address daiJoin_) public {
+        vat = VatLike(vat_);
+        daiJoin = DaiJoinLike(daiJoin_);
+        TokenLike(DaiJoinLike(daiJoin_).dai()).approve(daiJoin_, type(uint256).max);
+        VatLike(vat_).hope(daiJoin_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    uint256 internal constant RAY = 10 ** 27;
+
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "DssDirectDepositHub/underflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "DssDirectDepositHub/overflow");
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+    function _max(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x >= y ? x : y;
+    }
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _add(x, _sub(y, 1)) / y;
+    }
+
+    // --- Administration ---
+    /**
+        @notice update vow or end addresses.
+        @dev msg.sender must be authorized.
+        @param what name of what we are updating bytes32("vow"|"end")
+        @param data address we are setting it to
+    */
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+
+        if (what == "vow") vow = data;
+        else if (what == "end") end = EndLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    /**
+        @notice update tau value for D3M ilk.
+        @dev msg.sender must be authorized.
+        @param ilk  bytes32 of the D3M ilk to be updated
+        @param what bytes32("tau") or it will revert
+        @param data number of second to wait after caging a pool to write of debt
+    */
+    function file(bytes32 ilk, bytes32 what, uint256 data) external auth {
+        require(ilks[ilk].tic == 0, "DssDirectDepositHub/pool-not-live");
+
+        if (what == "tau" ) {
```

Comment:
Inconsistent use of braces: other `file` functions do not use them.

Reply:
1690c02